jt-classbreaks/src/main/java/it/geosolutions/jaiext/classbreaks/ClassPercentagesManager.java

```diff
@@ -0,0 +1,121 @@
+package it.geosolutions.jaiext.classbreaks;
+
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.function.Predicate;
+
+/**
+ * This class gathers method for percentages' computation for natural breaks,
+ * quantile and equal interval classifications
+ */
+public class ClassPercentagesManager {
+
+
+    //Quantile
+    public double[] getPercentages(Map<Double, Integer> data, List<Double> breaks, double nValues, int numClasses) {
+        double[] percentages = new double[numClasses];
+        Set<Double> keys = data.keySet();
+        for (int i = 0; i < numClasses; i++) {
+            double current = breaks.get(i);
+            double next = breaks.get(i + 1);
+            boolean last = numClasses == i + 1;
+            double classMembers;
+            if (last) {
+                classMembers = getClassMembersCount(data, keys, v -> v.doubleValue() >= current
+                        && v.doubleValue() <= next);
+            } else {
+                classMembers = getClassMembersCount(data, keys, v -> v.doubleValue() >= current
+                        && v.doubleValue() < next);
+            }
+            percentages[i] = (classMembers / nValues) * 100;
+        }
+        return percentages;
+
+    }
+
+    private int getClassMembersCount(Map<Double, Integer> data, Set<Double> keys, Predicate<Double> predicate) {
+        int classMembersCount = 0;
+        for (Double key : keys) {
+            if (predicate.test(key)) {
+                classMembersCount += data.get(key);
+            }
+        }
+        return classMembersCount;
+    }
+
+
+    // histogram natural breaks
+    public double[] getPercentages(List<HistogramClassification.Bucket> buckets,
+                                   Double[] breaks, int numClasses) {
+        double[] percentages = new double[numClasses];
+        int totalSize = 0;
+        for (int i = 0; i < buckets.size(); i++) {
+            totalSize += buckets.get(i).getCount();
+        }
+        int bucketSize = buckets.size();
+        for (int i = 0; i < numClasses; i++) {
+            double current = breaks[i];
+            double next = breaks[i + 1];
+            boolean last = numClasses == i + 1;
+            double classMembers;
+            if (last) {
```

Comment:
Given the tests checking alternatively the max, middle or min value, it does not seem impossible to find buckets that match both criteria, e.g, whose min is below the current break, but whose average sits in the next class.
Perhaps it is wrong to try and assign a bucket fully into a class, a different approach would be to assume the bucket is uniformly distributed, and assign counts to the output ranges based on the percentage of overlap instead?

Reply:
fixed